src/platform/x86_64/chal/chal_thd.h

```diff
@@ -0,0 +1,186 @@
+#ifndef CHAL_INVSTK
+#define CHAL_INVSTK
+
+/* code for hardware-specific thread routines */
+
+#include "component.h"
+#include "fpu_regs.h"
+
+
+#define THD_INVSTK_MAXSZ 32
+
+/*
+ * This is the data structure embedded in threads that are associated
+ * with an asynchronous receive end-point.  It tracks the reference
+ * count on that rcvcap, the tcap associated with it, and the thread
+ * associated with the rcvcap that receives notifications for this
+ * rcvcap.
+ */
+struct rcvcap_info {
+	/* how many other arcv end-points send notifications to this one? */
+	int            isbound, pending, refcnt, is_all_pending;
+	sched_tok_t    sched_count;
+	struct tcap *  rcvcap_tcap;      /* This rcvcap's tcap */
+	struct thread *rcvcap_thd_notif; /* The parent rcvcap thread for notifications */
+};
+
+typedef enum {
+	THD_STATE_PREEMPTED = 1,
+	THD_STATE_RCVING    = 1 << 1, /* report to parent rcvcap that we're receiving */
+} thd_state_t;
+
+struct invstk_entry {
+	struct comp_info comp_info;
+	unsigned long    sp, ip;
+	u8_t             ulinvstk_off;
+} HALF_CACHE_ALIGNED;
+
+/**
+ * The thread descriptor.  Contains all information pertaining to a
+ * thread including its registers, id, rcvcap information, and, most
+ * importantly, the kernel invocation stack of execution through
+ * components.
+ */
+struct thread {
```

Comment:
Everything we put in architecture-specific code, we have to *separately maintain* for each architecture. We want to avoid, where possible, replication between architectures for this reason. We can talk about ways to avoid most of this code in the architecture-specific code.